### PR TITLE
Use opaque metadata

### PR DESCRIPTION
**Pull request by frangio**
Fixes https://github.com/rugpullindex/ERC4973/issues/51

Please provide feedback on the naming.

### CODE REVIEW

File: src/interfaces/IERC4973.sol [hunk 1]
```diff
@@ -75,16 +75,21 @@ interface IERC4973 {
...
     address from,
-    string calldata uri,
+    bytes calldata metadata,
     bytes calldata signature
   ) external returns (uint256);
+  /// @notice Decodes the opaque metadata bytestring of an ABT into the token
+  /// URI that will be associated with it once it is created on chain.
+  /// @param metadata The metadata that will be associated to an ABT.
+  /// @return A URI that represents the metadata.
+  function decodeURI(bytes calldata metadata) external returns (string memory);
```
Comment (rahulrumalla) [inline src/interfaces/IERC4973.sol:94]:
```suggestion
  function decodeMetadata(bytes calldata metadata) external returns (string memory);
```

Would this naming be more appropriate?

Comment (frangio) [reply src/interfaces/IERC4973.sol:94]: `decodeURI` feels more accurate to me personally but I can see the ambiguity. We could scrap the word "decode" for something else too.

**Review by TimDaub — CHANGES_REQUESTED** [review]
please update the interface identifier everywhere too: https://github.com/rugpullindex/ERC4973/pull/52/files#diff-d6d069dfcb878b5b981b7d3dd18dce17746639eea3f30d1902f4a4cd46364f84R6

**Comment by TimDaub** [comment]
One question I have regarding the problem this is solving:

- Would this allow separating the tokenURI in e.g. baseURI and extension similar to how it was done with OpenZeppelin EIP-721 implementations?
- I'm reading your argument here: https://ethereum-magicians.org/t/eip-4973-account-bound-tokens/8825/149?u=timdaub and so my understand is that with this fix, a user e.g. wouldn't have to submit a base URI portion of the tokenURI etc. Is my understanding of this correct?

File: src/interfaces/IERC4973.sol (hunk 1 above)
Comment (TimDaub) [reply src/interfaces/IERC4973.sol:94]: I‘m ok with decodeURI

File: src/interfaces/IERC4973.sol [hunk 2]
```diff
@@ -75,16 +75,21 @@ interface IERC4973 {
...
```
Comment (TimDaub) [inline src/interfaces/IERC4973.sol:94]: Why are we including decodeURI in the interface? Shouldn‘t the interface user mainly interact with function tokenURI?

Comment (frangio) [reply src/interfaces/IERC4973.sol:94]: The problem is that `tokenURI` is only available after on-chain minting. The `metadata` value represents a URI but in an opaque way, and a wallet that holds that value wouldn't be able to  decode it into a URI, so the decoding function should be a part of the interface. You want to know the URI before you mint it.

Comment (TimDaub) [reply src/interfaces/IERC4973.sol:94]: reasonable point, I had never thought about it this way.

**Comment by TimDaub** [comment]
> * [ ]  please update the interface identifier everywhere too: https://github.com/rugpullindex/ERC4973/pull/52/files#diff-d6d069dfcb878b5b981b7d3dd18dce17746639eea3f30d1902f4a4cd46364f84R6
> * [ ]  Make foundry tests pass

I‘m happy to merge it when these points are completed

**Comment by TimDaub** [comment]
thx I‘ll take over getting this into the main branch. thanks again

**Comment by frangio** [comment]
@TimDaub Do you plan to merge this and also update the EIP?

**Comment by TimDaub** [comment]
Rebased and merged. Thank you again @frangio

**Pull request merged**